Pr ekf2 reset

tumbili opened the pull request.

LorenzMeier commented:
@tumbili Where is this standing?

tumbili commented:
@LorenzMeier @bkueng @julianoes This is tested in SITL and I flew it on a quad several times without issues. This would still need fixed wing testing. Review would be good!

julianoes commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_main.cpp:
```diff
@@ -2283,6 +2289,23 @@ FixedwingPositionControl::task_main()
+			// handle estimator reset events. we only adjust setpoins for manual modes
+			if (_control_mode.flag_control_manual_enabled) {
+				if (_control_mode.flag_control_altitude_enabled && _global_pos.alt_reset_counter != _alt_reset_counter) {
+					_hold_alt += _global_pos.delta_alt;
```
You are not resetting the `_alt_reset_counter`.

tumbili replied:
@julianoes That counter should not be reset! The ekf increased the counter after every reset and the position controller should track it.

julianoes replied:
Shouldn't there be this line here:
```
_alt_reset_counter = _global_pos.alt_reset_counter;
```

julianoes commented:
@tumbili Have you actually tested all the resets in SITL or flight tests, or have you just regression tested it?

tumbili commented:
@julianoes I have tested this in SITL by provoking a reset and making sure that the drone does not perform weird behavior. I have also flown this on a multicopter multiple times, however, I probably never had an actual reset occurring.

julianoes commented:
@tumbili do you think you could test actual resets? Otherwise this seems like code that goes in and never actually was verified.

LorenzMeier commented:
@tumbili What would be needed to test? A vehicle that can do logging be carried below a roof?

tumbili commented on src/modules/ekf2/ekf2_main.cpp:
```diff
@@ -807,12 +815,17 @@ void Ekf2::task_main()
 				global_pos.lat = est_lat; // Latitude in degrees
 				global_pos.lon = est_lon; // Longitude in degrees
+				map_projection_reproject(&ekf_origin, lpos.x - lpos.delta_xy[0], lpos.y - lpos.delta_xy[1], &lat_pre_reset, &lon_pre_reset);
+				global_pos.delta_lat_lon[0] = est_lat - lat_pre_reset;
```
@julianoes This probably needs wrapping, eh?

julianoes replied:
Yap, looks like it.

tumbili commented:
@LorenzMeier @julianoes We have 5 possible sources for a reset, namely position, height, vertical velocity, horizontal velocity an yaw. We can add some temporary code into ekf2 which adds offsets to the sensor data. We can use a switch to trigger it. We could have the following cases:

if aux1 > 0
1) Add constant 20 m to baro measurement
2) Add constant 10 m/s to gps down velocity
3) Add constant x, y lat lon offset to gps position measurement
4) Add constant 10 m/s to either gps vN or vE measurement
5) I need to see how we can trigger a yaw reset in the easiest way.

We could implement a temporary parameter which lets you switch between the different cases.
How does that sound?

tumbili commented:
@LorenzMeier @julianoes While testing this I found some bugs in ecl, this PR should fix it https://github.com/PX4/ecl/pull/199
I will discuss it with @priseborough so that we can get this in.

tumbili commented:
@julianoes @LorenzMeier Julian, I fixed up the stuff you commented on and added the lat lon coordinate wrap. Could you please review my  last two commits?
This is rebased on master and therefore, should have the ecl fixes.

tumbili commented:
@julianoes BTW, I also tested every possible reset in SITL except the yaw reset.
I still need to find a convenient way to trigger it.

julianoes commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_main.cpp:
```diff
@@ -2299,10 +2299,22 @@ FixedwingPositionControl::task_main()
+					_hdg_hold_prev_wp.lat = M_RAD_TO_DEG * asin(sin(_hdg_hold_prev_wp.lat * M_DEG_TO_RAD));
+
 					_hdg_hold_prev_wp.lon += _global_pos.delta_lat_lon[1];
+					_hdg_hold_prev_wp.lon = M_RAD_TO_DEG * matrix::wrap_pi(_hdg_hold_prev_wp.lon * M_DEG_TO_RAD);
```
I think matrix::wrap_pi maps to `[-180, 180)` (inclusive -180) while it probably should be `(-180, 180]` (inclusive 180) but that's really an edge case that the controller presumably handles.

julianoes commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_main.cpp:
```diff
@@ -2299,10 +2299,22 @@ FixedwingPositionControl::task_main()
+
+					// add position reset delta to previous waypoint coordinate
 					_hdg_hold_prev_wp.lat += _global_pos.delta_lat_lon[0];
+					_hdg_hold_prev_wp.lat = M_RAD_TO_DEG * asin(sin(_hdg_hold_prev_wp.lat * M_DEG_TO_RAD));
```
This seems like a costly operation, however, it's only used in case of a reset, so that should not be a problem.

But, please add a comment (and maybe even link to the stackoverflow page if you found it there). A future reader will be puzzled about what the hell this does :smile:.

tumbili replied:
@julianoes I knew you would say that :-p

priseborough commented:
@tumbili  The yaw reset triggers every time the EKF changes from magnetic yaw to  3D fusion. This happens after every takeoff when the hight has increased by 1.5m. If you takeoff from a steel surface or place a large steel object (large spanner, etc) underneath the copter before takeoff, then the change in yaw angle due to the reset will be more noticeable.

LorenzMeier commented:
@tumbili I did a rebase - can you address Julian's comment?

tumbili commented:
@LorenzMeier Done! So the only untested part is the reset for fixed wings. Maybe I could ask @CarlOlsson to test this on the plane during his weekly test flights. I would provide him with the code which adds glitches to gps data in order to provoke a reset. What do you think Carl?

CarlOlsson commented:
@tumbili of course, just send it over and a detailed test procedure and I will fly it!

tumbili commented:
@CarlOlsson Thanks! I pushed this branch to PX4 https://github.com/PX4/Firmware/tree/pr-ekf2ResetTest
which you can use for the test.

There is a param called "EKF2_TRIG_MODE"  with which you can trigger two different resets:

value = 0: no reset will be triggered

value = 1: this will suddenly subtract 20m from the baro sensor data. As a result the ekf will reject the baro measurements and you should get a height reset after 5 seconds. I suggest you either fly in ALTCTL or in POSCTRL. During the 5 seconds no absolute height measurements will be fused so the plane could drift a bit in height. However, after the reset it should hold it's current altitude.

value = 2: This will offset the gps position data. The effects should be more or less the same as described above. For this test you have to fly in POSCTRL. The plane should not suddenly change it's flight direction as a cause of the position reset. 
Don't forget that this modifies the gps data and so you can't rely on RTL working anymore :-)
Well, it will actually work, it will just land somewhere else.

So this is how it looks in code:
```
if (_reset_trigger.get() == 1) {
		sensors.baro_alt_meter -= 20.0f;

} else if (_reset_trigger.get() == 2) {
		gps.lat += 100000;	// add insane offset to positon, good luck!
		gps.lon += 100000;
}
```

Let me know if you have questions.

CarlOlsson commented:
Hi @tumbili here is the log file from today's filght:
http://logs.uaventure.com/view/VbRaqYkhoKEMU4YeyHFsDJ
I haven't had time to look if the resets actually happened, so just tell me if you wand it tested again

tumbili commented:
@CarlOlsson Thanks for the log, that helped much! I've found that we didn't make TECS handle the altitude resets so I did that as well.
I finally got my SITL plane setup running again and tested this branch in SITL.
It works as expected. I'll talk to you tomorrow about another test.

CarlOlsson commented:
I'm glad it was useful!

tumbili commented:
@CarlOlsson @LorenzMeier I tested this intensively in SITL and things work as expected.
Carl, if you could give this one more test flight on a plane that would be great!

LorenzMeier commented:
@tumbili Could you please fix the CI issues? Thanks!

tumbili commented:
Here are the SITL logs:
Plane: http://logs.uaventure.com/view/BUc4KLkbWaYjVwNVnxJ54h
Quad: http://logs.uaventure.com/view/H2ZGZaTHwUTVuHqWJJ7iqB

CarlOlsson commented:
@tumbili do you still want me to fly this? If so I can do it in a couple of hours

tumbili commented:
@CarlOlsson I pushed the pr-ekf2ResetTest branch, all good to go. I now also includes the recent changes of ecl.

tumbili commented:
@CarlOlsson Have you managed to do some testing on a fw? If this is too much of a pain we should get a SITL test done.

CarlOlsson commented:
No I have not been out yet, but will go tomorrow!

tumbili commented:
@LorenzMeier @priseborough @CarlOlsson This is rebased on master and tested in SITL.
I added offsets to baro and gps measurement to trigger ekf resets. I triggered the resets in both manual and automated modes and the results were always as expected.
In manual modes the corresponding setpoints were adjusted such that the reset was not even noticeable.
I mission modes the setpoints were not shifted and the drone reacted correctly.
I tested for both multicopter and fixed wing, everything worked as expected.
Here are two logs for reference, the first one are the results for the multicopter, the second one for fixed wing.

Multicopter
http://logs.uaventure.com/view/KAoPE7F8FZTK653CWi4ouE

Fixed wing
http://logs.uaventure.com/view/vj7CvHnJDtFiKp7gRbXLwM

LorenzMeier commented:
Looks good to merge for me.

tumbili commented:
Merged
